plugin-server/src/utils/event.ts

```diff
@@ -111,6 +96,27 @@ export function convertToIngestionEvent(event: RawClickHouseEvent, skipElementsC
     }
 }
 
+/**
+ * Elements parsing can be really slow so it is only done when required by the caller.
+ * It mutates the event which is not ideal but the performance gains of lazy loading it were deemed worth it.
+ */
+export function mutatePostIngestionEventWithElementsList(event: PostIngestionEvent): void {
+    if (event.elementsList) {
+        // Don't set if already done before
+        return
+    }
+
+    event.elementsList = event.properties['$elements_chain']
```

Comment:
test failures: `event.properties` can be `undefined`

Reply:
Ooh then why is the type concrete? Should I change that?

Reply:
good point it's Record here https://github.com/PostHog/plugin-scaffold/blob/main/src/types.ts#L68 so it would be better to fix in the tests where we create mocks badly?

Reply:
Hmmm lots of nasty `as any` casting in the tests. Can this ever _actually_ be undefined?

Reply:
I don't know 🤷‍♀️  we might have as in other places too - so just doing `(event.properties || {})[$elements_chain]` here would be the simplest path forward

Reply:
Yeah but its smelly. If it can be undefined - we should make it undefined. Otherwise we don't have any way of catching it. I'll fix the tests rather